[VPlan] Recursively match operands of interleave group

artagnon opened the pull request:
When narrowing interleave groups, we use the canNarrowLoad check, which bails out when there any recipe that is not a VPWidenLoad, VPWidenInterleave, or live-in feeding the interleave: a lot of potential narrowing opportunities are missed as a result. Fix this.

Fixes #128062.

artagnon commented on llvm/test/Transforms/LoopVectorize/pr128062-interleaved-accesses-narrow-group.ll:
```diff
@@ -0,0 +1,182 @@
+; CHECK-NEXT:    [[INDEX:%.*]] = phi i64 [ 0, %[[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], %[[VECTOR_BODY]] ]
+; CHECK-NEXT:    [[OFFSET_IDX:%.*]] = mul i64 [[INDEX]], 4
+; CHECK-NEXT:    [[NEXT_GEP:%.*]] = getelementptr i8, ptr [[DST_START]], i64 [[OFFSET_IDX]]
+; CHECK-NEXT:    [[WIDE_VEC:%.*]] = load <16 x i8>, ptr [[NEXT_GEP]], align 1
```
Thanks for spotting! I think the regression cases also do this: investigating.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/transform-narrow-interleave-to-widen-memory-constant-ops.ll:
```diff
@@ -175,33 +178,24 @@ define void @test_add_double_same_var_args_1(ptr %res, ptr noalias %A, ptr noali
-; CHECK-NEXT:    [[INTERLEAVED_VEC5:%.*]] = shufflevector <4 x double> [[TMP10]], <4 x double> poison, <4 x i32> <i32 0, i32 2, i32 1, i32 3>
-; CHECK-NEXT:    store <4 x double> [[INTERLEAVED_VEC5]], ptr [[TMP8]], align 4
-; CHECK-NEXT:    [[INDEX_NEXT]] = add nuw i64 [[INDEX]], 4
+; CHECK-NEXT:    store <2 x double> [[TMP5]], ptr [[TMP7]], align 4
```
I am not 100% sure why we did not cover this before with the old logic?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -4108,25 +4116,51 @@ VPlanTransforms::expandSCEVs(VPlan &Plan, ScalarEvolution &SE) {
+/// Returns true if the \p StoredValues of an interleave group match. It does
+/// this by going through operands recursively until it hits the leaf cases:
+/// VPWidenLoadRecipe, VPInterleaveRecipe, and live-ins.
+static bool interleaveStoredValuesMatch(ArrayRef<VPValue *> StoredValues) {
```
I think we should be able to keep the current recursive structure and still generalize it to catch the motivating cases?

It seems the current PR tries to generalize along multiple dimensions, but it would probably be good to do step by-step.

Support chains of wide recipes could be done as in https://github.com/llvm/llvm-project/pull/167309 + https://github.com/llvm/llvm-project/pull/167310

artagnon commented:
I think this is the only viable approach to achieving full generality.

artagnon committed:
[VPlan] Recursively match operands of interleave group

When narrowing interleave groups, we use the canNarrowLoad check, which
bails out when there any recipe that is not a VPWidenLoad,
VPWidenInterleave, or live-in feeding the interleave: a lot of potential
narrowing opportunities are missed as a result. Fix this.

Fixes #128062.

artagnon commented:
This patch has been re-done completely, and is now ready for review.

eas commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -5130,25 +5130,116 @@ VPlanTransforms::expandSCEVs(VPlan &Plan, ScalarEvolution &SE) {
+  for (unsigned I = 0; I != Operands.size(); ++I) {
+    if (!Operands[I]->hasDefiningRecipe())
+      continue;
+    VPRecipeBase *Cur = Operands[I]->getDefiningRecipe();
```
I misread the type...

fhahn reviewed (commented):
It looks like this is causing a number of crashes, including for https://llvm.godbolt.org/z/PKGEK7jxr

On the high-level, it would be very helpful to have the new code documented more, to make it easier to understand what's going on. As-is it is a bit hard to follow/verify all cases are handled correctly.

Does this handle cases we can't get with just doing plain recursive matching, as in https://github.com/llvm/llvm-project/pull/167310 + extending that to handle more known safe recipes?
